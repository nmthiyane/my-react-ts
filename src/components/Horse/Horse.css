.HorseDiv{
    position: relative;
    text-align: left;
    padding: 20px;
    padding-bottom: 0px;
    padding-top: 0px;
}

.HorseDiv h4{
    bottom: 10px;
}

.HorseDiv progress{
    position: relative;
    height: 60px;
    text-align: center; /* To center it horizontally (if you want) */
    line-height: 30px; /* To center it vertically */
    width: 90% ;
    color: blueviolet; 
    margin: auto;
    padding-bottom: 20px;
    padding-left: 30px;
}

.HorseDiv img{
    height: 60px;
    width: 60px;
    padding-top: 0px;
}
