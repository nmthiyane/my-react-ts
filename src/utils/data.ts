import { Racer } from '../models';

const data: Racer[] = require('../data/team.json');

/**
 * Returns an array of racers
 *
 * @export
 * @returns {Racer[]}
 */
export function getRacerData(): Racer[] {
	return data;
}