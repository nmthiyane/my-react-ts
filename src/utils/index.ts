export * from './race';
export * from './data';