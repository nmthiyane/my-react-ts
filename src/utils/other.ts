/**
 * Checks to see if the value is null or udefined
 *
 * @export
 * @param {*} value Value to check
 * @returns {boolean} true if it's null or udefined else false
 */
export function isNullOrUndefined<T>(value: T | null | undefined): value is null | undefined {
	return value === null || isUndefined(value);
}

/**
 * Will validate the passed parameter against isNullOrUndefined, isObjectAndEmpty, isArrayAndEmpty, isStringAndEmpty
 *
 * @export
 * @param {(any[] | undefined)} collection
 * @returns {boolean}
 */
export function isNullOrEmpty(parameter: any[]): boolean {
	return isNullOrUndefined(parameter) || isStringAndEmpty(parameter) || isArrayAndEmpty(parameter) || isObjectAndEmpty(parameter);
}

/**
 * Checks if a value is undefined
 *
 * @export
 * @template T
 * @param {(T | undefined)} value
 * @returns {value is undefined}
 */
export function isUndefined<T>(value: T | undefined): value is undefined {
	return value === void 0;
}

/**
 * Checks if a string is null or whitespace.
 *
 * @export
 * @param {string} value to check if the value is null or whitespace
 * @returns {boolean} true if it's null, undefined, empty or whitespace else false
 */
export function isNullOrWhitespace(value: string): boolean {
	return (value || '').trim() === '';
}

/**
 * Validates if the parameter is of type string and if it is null or whitespace
 *
 * @export
 * @param {any} value
 * @returns {boolean}
 */
export function isStringAndEmpty(value: any): boolean {
	return typeof value === 'string' && isNullOrWhitespace(value);
}

/**
 * Will validate if the parameter is of type array and if it is empty
 *
 * @export
 * @param {any} array
 * @returns {boolean}
 */
export function isArrayAndEmpty(arr: any): boolean {
	return !isNullOrUndefined(arr) && Array.isArray(arr) && arr.length === 0;
}

/**
 * Will return a boolean indicating wether the passed variable is an object
 *
 * @export
 * @param {any} obj
 * @returns {boolean}
 */
export function isObject(obj: any): boolean {
	return (typeof obj === 'object' && !Array.isArray(obj));
}

/**
 * Will validate if the passed parameter is of type object and has no properties defined
 *
 * @param {{}} obj
 * @returns {boolean}
 */
export function isObjectAndEmpty(obj: {}): boolean {
	if (!isObject(obj)) {
		return false;
	}

	for (let prop in obj) {
		if (obj.hasOwnProperty(prop)) {
			return false;
		}
	}

	return JSON.stringify(obj) === JSON.stringify({});
}

/**
 * Does a deepcopy of a multi level object
 * Object.assign does not copy items that are lower than the root level
 * those items will be copied by reference
 *
 * @export
 * @template T
 * @param {(T | undefined)} [value]
 * @returns {(T | null)}
 */
export function deepCopy<T>(value?: T | undefined): T | null {
	if (!isUndefined(value)) {
		let response = JSON.parse(JSON.stringify(value));
		return response;
	}
	return null;
}