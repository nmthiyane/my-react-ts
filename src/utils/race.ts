/**
 * Calculates the next posistion of a racer using the current position
 *
 * @export
 * @param {number} currentPosition
 * @returns {number}
 */
export function calculateRacerPosition(currentPosition: number): number {
	return currentPosition + Math.max(Math.random() * 10);
}