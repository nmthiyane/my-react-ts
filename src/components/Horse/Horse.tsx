import * as React from 'react';
import './Horse.css';

import { Racer } from '../../models';
import { calculateRacerPosition } from '../../utils';

interface Props {
	racer: Racer;
	trackOffset?: Function;
	raceInProgress?: boolean;
	raceStartPosition?: boolean;
	setWinner?: (winner: string) => void;
}

interface State {
	racerPosition: number;
}

export class Horse extends React.Component<Props, State> {
	constructor(props: Props) {
		super(props);

		this.state = {
			racerPosition: 0
		};

		this.determineRacerPosition = this.determineRacerPosition.bind(this);
		this.doRacing = this.doRacing.bind(this);
	}

	componentWillUpdate(nextProps: Props, nextState: State) {
		const { racerPosition } = nextState;
		const { raceInProgress, setWinner, racer, trackOffset, raceStartPosition } = nextProps;
		if (raceInProgress) {
			if ((trackOffset && trackOffset()) - 100 > racerPosition) {
				this.doRacing();
			}
			else if (setWinner) {
				setWinner(racer.login);
			}
		}
		else if (raceStartPosition && racerPosition !== 0) {
			this.setState({ racerPosition: 0 });
		}
	}

	determineRacerPosition() {
		const { racerPosition } = this.state;
		this.setState({ racerPosition: calculateRacerPosition(racerPosition) });
	}

	doRacing() {
		setTimeout(() => this.determineRacerPosition(), 15);
	}

	render() {
		const { racer } = this.props;
		const { racerPosition } = this.state;
		return (
			<div
				className="Horse"
				style={{ paddingLeft: `${racerPosition}px` }}
			>
				<img
					height="100px"
					src={racer.avatar_url}
				/>
			</div>
		);
	}
}