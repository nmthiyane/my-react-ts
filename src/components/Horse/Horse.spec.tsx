import * as React  from 'react';
import { Horse } from './Horse';
import { Racer } from '../../models';
import { getRacerData } from '../../utils';

const renderer  = require('react-test-renderer');

describe('Snapshot test for the Horse component', () => {
	
	let racer: Racer = getRacerData()[0];
	it('renders a snapshot', () => {
	const tree = renderer.create((
		<Horse 
			racer={racer}
		/>
		)).toJSON();
	expect(tree).toMatchSnapshot();
	});
});