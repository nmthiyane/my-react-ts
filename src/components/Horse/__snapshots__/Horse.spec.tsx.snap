// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for the Horse component renders a snapshot 1`] = `
<div
  className="Horse"
  style={
    Object {
      "paddingLeft": "0px",
    }
  }
>
  <img
    height="100px"
    src="https://avatars2.githubusercontent.com/u/1482046?v=4"
  />
</div>
`;
